cpp/src/arrow/compute/exec.cc

```diff
@@ -697,7 +771,18 @@ class ScalarExecutor : public KernelExecutorImpl<ScalarKernel> {
       }
     }
 
+    // To check whether the kernel allocated new Buffers,
+    // insert all the preallocated ones into a set
+    std::unordered_set<std::shared_ptr<Buffer>> pre_buffers;
```

Comment:
I'm not sure how we want to handle this feature at all, but we definitely don't want to run those checks unconditionally, unless they're extremely trivial. Perhaps this can be enabled only on debug builds? (`#ifndef _NDEBUG`, for example)

cc @lidavidm

Reply:
I think debug-only is fine as this will be most useful for developers.

Reply:
The checks are now inside `#ifndef NDEBUG`.